### PR TITLE
Starting version of speech to text robots

### PR DESCRIPTION
**Pull request by peetucket**
## Why was this change made? 🤔

Part of #1341 

1. Renames "caption" robots to "speechToText" robots (workflow definition itself in workflow-server-rails still to come)
2. Adds likely candidates for other new robots will we need (analogous to ocrWF)
3. Create new class to handle opening an object with retries since we will want to do this for both ocr and speech to text
4. Use this new class in existing ocr robots and new speechToText robots to open objects.
5. Create a new class which will be used to determine if speech to text is needed (analogous to existing class for OCR). 
6. New tests

Workflow definition in https://github.com/sul-dlss/workflow-server-rails/pull/804

## How was this change tested? 🤨

Unit tests

### CODE REVIEW

File: lib/dor/text_extraction/ocr.rb [hunk 1]
```diff
@@ -127,7 +127,7 @@ def cleanup_abbyy_exceptions
         end
       end
 
-      # iterate through cocina strutural contains and return all File objects for files that need to be OCRed
+      # iterate through cocina structural contains and return all File objects for files that need to be OCRed
```
Comment (peetucket) [inline lib/dor/text_extraction/ocr.rb:130]: typo

File: lib/dor/text_extraction/ocr.rb [hunk 2]
```diff
@@ -160,8 +160,7 @@ def allowed_mimetypes
         ]
       end
 
-      # maps the allowed content types to the resource type we will look for files in
-      # TODO: refine list of allowed object types for OCR
+      # maps the allowed object types to the resource type we will look for files in
```
Comment (peetucket) [inline lib/dor/text_extraction/ocr.rb:163]: remove the 'todo' for OCR ... we are done for now

File: lib/dor/text_extraction/version_updater.rb [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+          raise e unless tries < max_tries
+
+          Honeybadger.notify('[NOTE] Problem opening object version', context: { description:, druid:, tries:, error: e })
+          retry
+        end
+      end
+      # rubocop:enable Metrics/MethodLength
+    end
+  end
+end
```
Comment (peetucket) [inline lib/dor/text_extraction/version_updater.rb:41]: new class to handle object opening with retries (extracted from start-ocr robot so we can use in start-stt robot)

File: lib/dor/text_extraction/speech_to_text.rb [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+      def matchtype
+        resource_type_mapping[cocina_object.type]
+      end
+
+      def allowed_object_types
+        resource_type_mapping.keys
+      end
+    end
+  end
+end
```
Comment (peetucket) [inline lib/dor/text_extraction/speech_to_text.rb:95]: analogous to ocr class to handle various speech to text related tasks (put in some sensible defaults, we can edit later on implementation)

File: spec/robots/dor_repo/ocr/start_ocr_spec.rb [hunk 5]
```diff
@@ -36,41 +36,6 @@
...
-      before { allow(version_client).to receive(:open).and_raise(Dor::Services::Client::UnexpectedResponse.new(response: 'nope')) }
-
-      it 'logs to honeybadger and then raises the error' do
-        expect { perform }.to raise_error(Dor::Services::Client::UnexpectedResponse)
-        expect(version_client).to have_received(:open).thrice # shakespeare coding (all three times fail)
-        expect(Honeybadger).to have_received(:notify).twice # two calls to HB for the first two failures
-      end
-    end
-  end
-
```
Comment (peetucket) [inline spec/robots/dor_repo/ocr/start_ocr_spec.rb:73]: these tests are no longer needed here and are instead performed in the new spec for the new `VersionUpdated` class

**Review by edsu — APPROVED** [review]
This all looks good. I left a comment that the approach to mime type matching may need to be generalized. I found myself kind of wanting to expand `stt` to `speech_to_text` and `SpeechToText` in places, but I can live with it being abbreviated.

File: lib/dor/text_extraction/speech_to_text.rb [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        perservedfiles = fileset.structural.contains.select { |file| file.administrative.sdrPreserve && allowed_mimetypes.include?(file.hasMimeType) }
+        return perservedfiles[0] if perservedfiles.one?
+
+        perservedfiles = perservedfiles.sort_by { |pfile| allowed_mimetypes.index(pfile.hasMimeType) }
+        perservedfiles[0]
+      end
+
+      # defines the mimetypes types for which speech to text files can possibly be run
+      # preferentially select files for speech to text by ordering of mimetypes below
+      # TODO: refine list of allowed mimetypes for speech to text
```
[the remark this answers is not in the document]
Comment (peetucket) [reply lib/dor/text_extraction/speech_to_text.rb:68]: Yeah, good question.  I'd suggest we adjust later when we know better that types of formats the sul-embed media player will want.  I took this starter set from actual media objets, so they are at least somewhat real.

**Comment by peetucket** [comment]
> This all looks good. I left a comment that the approach to mime type matching may need to be generalized. I found myself kind of wanting to expand `stt` to `speech_to_text` and `SpeechToText` in places, but I can live with it being abbreviated.
> 
> I was wondering if it might be worth doing an integration test before merging?

Deploying this and running the OCR integration tests is a good idea.  I will do this before merging.

I am not ambivalent about the "stt" stuff - that's in the individual steps within the workflow itself, so it's was mostly to save space (it should be more clear what we are talking about given the workflow name is fully descriptive)

File: lib/dor/text_extraction/speech_to_text.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (jmartin-sul) [reply lib/dor/text_extraction/speech_to_text.rb:68]: i'd be fine with adjusting later.  maybe worth a backlog ticket for coming back to re-examine later?  or will this be obvious enough for us to not forget?

**Comment by jmartin-sul** [comment]
> > This all looks good. I left a comment that the approach to mime type matching may need to be generalized. I found myself kind of wanting to expand `stt` to `speech_to_text` and `SpeechToText` in places, but I can live with it being abbreviated.
> > I was wondering if it might be worth doing an integration test before merging?
> 
> Deploying this and running the OCR integration tests is a good idea. I will do this before merging.
> 
> I am not ambivalent about the "stt" stuff - that's in the individual steps within the workflow itself, so it's was mostly to save space (it should be more clear what we are talking about given the workflow name is fully descriptive)

i also found myself sort of wanting `stt` expanded to prevent future confusion, but it's a good point that all the uses occur inside of `SpeechToText` (for the robot code) or `speechToTextWF` (for the workflow xml).  so i think that makes me neutral on that point.  either way, if the change were made, i'd lean slightly towards doing that in a follow-on PR, since it'd be easier to review separately.  or at least as a separate commit (given that it'd affect most or all of the files in this PR, which is necessarily large, and a bit unweildy to review).

**Review by jmartin-sul — APPROVED** [review]
i had a small number of nitpicky questions/suggestions, but i think i'd rather just merge this and do those in a follow up PR, if you think they're worthwhile.

agreed that an integration test before merging is a good idea, given the refactor of existing functionality (though the refactor LGTM based just on reading through).

File: spec/robots/dor_repo/speech_to_text/start_stt_spec.rb [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+  context 'when the object is not opened and it is not possible to run text extraction' do
+    let(:version_open) { false }
+    let(:possible) { false }
+    let(:note) { 'No files available or invalid object for Speech To Text' }
+    let(:workflow) { 'speechToTextWF' }
+    let(:status) { 'skipped' }
+    let(:workflow_client) { instance_double(Dor::Workflow::Client) }
+    # NOTE: this is just mocking a workflow response, it doesn't actually have to be fully accurate
+    # we just need to expect we will update the status of all steps except start-stt
+    let(:skip_all) { instance_double(Dor::Workflow::Response, name: 'skip-all') }
```
[the remark this answers is not in the document]
Comment (peetucket) [reply spec/robots/dor_repo/speech_to_text/start_stt_spec.rb:48]: it's actually used in a `before` block just below

File: lib/dor/text_extraction/speech_to_text.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (peetucket) [reply lib/dor/text_extraction/speech_to_text.rb:68]: Creating a ticket so we don't forget.

**Comment by peetucket** [comment]
> i also found myself sort of wanting `stt` expanded to prevent future confusion, but it's a good point that all the uses occur inside of `SpeechToText` (for the robot code) or `speechToTextWF` (for the workflow xml). so i think that makes me neutral on that point. either way, if the change were made, i'd lean slightly towards doing that in a follow-on PR, since it'd be easier to review separately. or at least as a separate commit (given that it'd affect most or all of the files in this PR, which is necessarily large, and a bit unweildy to review).

Integration test done and works fine for OCR.

Will leave any potential renaming from `stt` to `speech_to_text` to a future PR.

File: spec/lib/dor/text_extraction/speech_to_text_spec.rb [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+  def build_file(sdr_peserve, filename)
+    extension = File.extname(filename)
+    mimetype = { '.wav' => 'audio/x-wav', '.mp4' => 'video/mp4', '.txt' => 'text/plain' }
+    sdr_value = instance_double(Cocina::Models::FileAdministrative, sdrPreserve: sdr_peserve)
+    instance_double(Cocina::Models::File, administrative: sdr_value, hasMimeType: mimetype[extension], filename:)
+  end
+
+  describe '#possible?' do
+    context 'when the object is not a DRO' do
+      let(:cocina_object) { instance_double(Cocina::Models::DRO, externalIdentifier: druid, dro?: false, type: object_type) }
```
[the remark this answers is not in the document]
Comment (peetucket) [reply spec/lib/dor/text_extraction/speech_to_text_spec.rb:28]: First off, thank for the Simpsons reference (actually had to look that one up and I'm a Simpsons fan!).  And yes, making the double a `Collection` is less confusing and works just fine.  I'll change here and in the similar OCR test.

File: lib/dor/text_extraction/version_updater.rb [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+      attr_reader :druid, :object_client, :description, :max_tries
+
+      def initialize(druid:, object_client:, description:, max_tries:)
+        @druid = druid
+        @object_client = object_client
+        @description = description
+        @max_tries = max_tries
+      end
+
+      def self.open(druid:, object_client:, description: '', max_tries: 3)
```
[the remark this answers is not in the document]
Comment (peetucket) [reply lib/dor/text_extraction/version_updater.rb:16]: Fair enough.  We only have a couple callers and they provide descriptions, so no need for a default.

**Comment by peetucket** [comment]
I believe I addressed all of the comments (except for the potential renaming, which is left to future work).  Will wait a bit to see if anyone wants to re-check, and will then merge if no objections.

**Pull request merged**